### ISSUE TITLE
Clean up errors

### ISSUE THREAD

**Issue by neachdainn** [adeschamps/lcm-rust#13]
## Current state of this issue

There are a significant number of `ProviderIssue` type errors that do not give the end user any information and requires that they look at the log to determine what went wrong. Trying to include either `#[cause] Box<failure::Fail>` or `#[cause] failure::Error` do not work without additional infrastructure [[1]](https://github.com/adeschamps/lcm-rust/pull/33#issuecomment-369114919). Using `#[cause] failure::Compat<failure::Error>` will compile but will not work as expected [[2]](https://github.com/adeschamps/lcm-rust/pull/33#issuecomment-369362884).

An [Error and ErrorKind pair](https://boats.gitlab.io/failure/error-errorkind.html) has been suggested but the current plan is to wait until this library has been used some and until `failure` hits 1.0 [[3]](https://github.com/adeschamps/lcm-rust/pull/33#issuecomment-369462132).

## Original Post

Right now, the error situation is a bit of a mess. This is my first time using `failure` and I haphazardly added error types as I went. This definitely needs to be cleaned up to provide a more informative and usable interface.

At the very least, I think the `Lcm` methods should all return a single error. What this error would look like, I don't know. But the major problem right now is that having them all return different types makes using them together painful without using the `failure::Error` type (which is not an ideal type to use).

In other words, the bare minimum is that we need to find an error type `T` such that the following code compiles:
```rust
fn do_stuff() -> Result<(), T> {
    let mut lcm = Lcm::new()?;
    let sub = lcm.subscribe("TEST", |m: Foo| { .. })?;
    lcm.publish("OTHER TEST", m)?;

    lcm.handle()?;
    Ok(())
}
```

**Comment by neachdainn** [comment]
@adeschamps 

Most of the other issues I am submitting right now are just things that need to be done. I think this one actual deserves some discussion. I also think that we really need to resolve this before the crate is published.

**Comment by adeschamps** [comment]
So, right now there are six error types:
- DecodeError
- EncodeError
- HandleError
- LcmInitError
- PublishError
- SubscribeError

I think it's important to think about what one could reasonably do in response to each of these errors.

If the best response is to print or log a message of some sort, then that's a good use case for the `failure::Error` type. In your example, I think `failure::Error` is actually the appropriate type - `do_stuff()` could fail for so many reasons, and the caller is unlikely to do anything other than log something or pass the error on.

There may be some places where the library should uphold some invariants that make it safe to unwrap errors. For example, we `impl From<io::Error> for EncodeError` because the `Write` trait has methods that return `Result`s - and while that's applicable to files and sockets, in our when we encode messages we're just writing to a `Vec`, so it should never fail. So we can probably eliminate a few types and variants as well.

**Comment by neachdainn** [comment]
> the caller is unlikely to do anything other than log something or pass the error on.

I agree that logging the error and moving on is the most likely response to an error. However, I am hesitant to just return `failure::Error` based on [this Reddit post and response](https://www.reddit.com/r/rust/comments/7te8si/personal_experience_can_cause_hard_to_notice/dtcdp0m/) and because the impression I've gotten is that the type is meant for binaries and prototyping. That said, a single error type that may not be as informative is much, much better than six.

Here are the current errors listed out, so it'll be easier to discuss:
* `LcmInitError`
    * Provider failed to start (cause: `io::Error`)
    * Unknown provider
    * Invalid LCM URL
* `SubscribeError`
    * Invalid regex (cause: `regex::Error`)
    * Provider failed/shutdown
* `PublishError`
    * Failed to encode message (cause: `EncodeError`)
    * Provider had IO issues (cause: `io::Error`)
    * Message wasn't sent
    * Message was too large to send
    * Channel name was too long
* `HandleError`
    * Provider had IO issues (cause: `io::Error`)
    * Provider failed/shutdown
* `EncodeError`
    * Size of array doesn't match the specified size.
    * Error reading from buffer (cause: `io::Error`) <sup>1</sup>
* `DecodeError`<sup>2</sup>
    * Invalid array size
    * Hash didn't match
    * String wasn't UTF-8 (cause: `string::FromUtf8Error`)
    * Invalid boolean
    * Missing null terminator on string
    * SPSC for channel was closed <sup>3</sup>
    * Error writing to buffer (cause: `io::Error`) <sup>1</sup>

1: Probably not needed, but it lets us use `?`.
2: Not immediately exposed to the user. They user can see them via the `Marshall` trait, but really only used in the provider code.
3: Signifies an unsubscribe event.

**Comment by neachdainn** [comment]
I think the major question is: which of these can people respond to in code and which should just log an error and return a generic message?
